open_spiel/games/2048.cc

```diff
@@ -0,0 +1,382 @@
+        return Coordinate(-1, 0);        
+      case kMoveRight:
+        return Coordinate(0, 1);
+      case kMoveDown:
+        return Coordinate(1, 0);
+      case kMoveLeft:
+        return Coordinate(0, -1);
+    }
+}
+
+std::vector<Coordinate> TwoZeroFourEightState
+    ::FindFarthestPosition(int r, int c, int direction) const {  
+  // Progress towards the vector direction until an obstacle is found
+  Coordinate prev = Coordinate(r, c);
+  do {
+    prev = Coordinate(r, c);
+    Coordinate direction_diff = GetVector(direction);
+    r += direction_diff.row;
+    c += direction_diff.column;
+  } while (WithinBounds(r, c) && CellAvailable(r, c));
+  return std::vector<Coordinate> {prev,
+      Coordinate(r, c)};
+};
+
+// Check for available matches between tiles (more expensive check)
+bool TwoZeroFourEightState::TileMatchesAvailable() const {
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      int tile = BoardAt(r, c).value;
+      if (tile > 0) {
+        for (int direction = 0; direction < 4; direction++) {
+          Coordinate vector = GetVector(direction);
+          int other = GetCellContent(r + vector.row, c + vector.column);
+          if (other > 0 && other == tile) {
+            return true; // These two tiles can be merged
+          }
+        }
+      }
+    }
+  }
+  return false;
+};
+
+void TwoZeroFourEightState::PrepareTiles() {
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      Tile tile = BoardAt(r, c);
+      if (tile.is_merged) {
+        SetBoard(r, c, Tile(tile.value, false));
+      }
+    }
+  }  
+};
+
+int TwoZeroFourEightState::GetCellContent(int r, int c) const {
+  if (!WithinBounds(r, c))
+    return 0;
+  return BoardAt(r, c).value;
+}
+
+void TwoZeroFourEightState::DoApplyAction(Action action) {
+  if (IsChanceNode()) {
+    // The original 2048 game starts with two random tiles
+    if (!extra_chance_turn_) {
+      current_player_ = 0;
+    }
+    extra_chance_turn_ = false;
+    
+    if (action == kNoCellAvailableAction) {
+      return;
+    }
+    ChanceAction chance_action = SpielActionToChanceAction(action);
+    SetBoard(chance_action.row, chance_action.column,
+        Tile(chance_action.is_four ? kChanceTiles[1] : kChanceTiles[0], false));
+    return;
+  }
+  action_score_ = 0;
+  std::vector<std::vector<int>> traversals = BuildTraversals(action);
+  PrepareTiles();
+  for (int r : traversals[0]) {
+    for (int c : traversals[1]) {
+      int tile = GetCellContent(r, c);
+      if (tile > 0) {
+        bool moved = false;
+        std::vector<Coordinate> positions = FindFarthestPosition(r, c, action);
+        Coordinate farthest_pos = positions[0];
+        Coordinate next_pos = positions[1];
+        int next_cell = GetCellContent(next_pos.row, next_pos.column);
+        if (next_cell > 0 && next_cell == tile
+            && !BoardAt(next_pos.row, next_pos.column).is_merged) {
```

Comment:
This would be cleaner if BoardAt took a Coordinate.

Reply:
Changed this